src/json.ts

```diff
@@ -34,6 +40,109 @@ export type Json =
+ * type Foo = ObjectOptional<{ foo: string | undefined }>;
+ * // Foo is equivalent to { foo?: string }
+ * ```
+ */
+export type ObjectOptional<Schema extends Record<string, unknown>> = {
+  [Key in keyof Schema as Schema[Key] extends ExactOptionalGuard
+    ? Key
+    : never]?: Schema[Key] extends ExactOptionalGuard & infer Original
+    ? Original
+    : never;
+} & {
+  [Key in keyof Schema as Schema[Key] extends ExactOptionalGuard
+    ? never
+    : Key]: Schema[Key];
+};
+
+/**
+ * An object type with support for exact optionals. This is used by the `object`
+ * struct. This uses the {@link ObjectOptional} helper to make properties with
+ * `undefined` in their type optional, but not `undefined` itself.
+ */
+export type ObjectType<Schema extends ObjectSchema> = Simplify<
+  ObjectOptional<
+    Optionalize<{
+      [Key in keyof Schema]: Infer<Schema[Key]>;
+    }>
+  >
+>;
+
+/**
+ * A struct to check if the given value is a valid object, with support for
+ * {@link exactOptional} types.
+ *
+ * @param schema - The schema of the object.
+ * @returns A struct to check if the given value is an object.
+ */
+export const object = <Schema extends ObjectSchema>(
+  schema: Schema,
+): Struct<ObjectType<Schema>> =>
+  // The type is slightly different from a regular object struct, because we
+  // want to make properties with `undefined` in their type optional, but not
+  // `undefined` itself. This means that we need a type cast.
+  superstructObject(schema) as unknown as Struct<ObjectType<Schema>>;
+
+declare const exactOptionalSymbol: unique symbol;
+type ExactOptionalGuard = {
+  _exactOptionalGuard?: typeof exactOptionalSymbol;
+};
+
+/**
+ * A struct to check if the given value is valid, or not present. This means
+ * that it allows an object which does not have the property, or an object which
+ * has the property and is valid, but not an object which has the property set
+ * to `undefined`.
+ *
+ * This struct should be used in conjunction with the {@link object} from this
+ * library, to get proper type inference.
+ *
+ * @param struct - The struct to check the value against, if present.
+ * @returns A struct to check if the given value is valid, or not present.
+ * @example
+ * ```ts
+ * const struct = object({
+ *   foo: exactOptional(string()),
+ *   bar: exactOptional(number()),
+ *   baz: optional(boolean()),
+ *   qux: unknown(),
+ * });
+ *
+ * type Type = Infer<typeof struct>;
+ * // Type is equivalent to:
+ * // {
+ * //   foo?: string;
+ * //   bar?: number;
+ * //   baz?: boolean | undefined;
+ * //   qux: unknown;
+ * // }
+ * ```
+ */
+export const exactOptional = <Type, Schema>(
+  struct: Struct<Type, Schema>,
+): Struct<Type & ExactOptionalGuard, null> =>
+  define('optional', (value, context) => {
+    const parent = context.branch[context.branch.length - 2];
+    const key = context.path[context.path.length - 1];
+
+    if (!hasProperty(parent, key)) {
+      return true;
+    }
+
+    if (value === undefined) {
```

Comment:
I don't think this check is required, it should be left to `struct.validator` to verify.